service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/aws/sageMakerNotebook/StopAwsSageMakerNotebookStep.java

```diff
@@ -0,0 +1,118 @@
+package bio.terra.workspace.service.resource.controlled.cloud.aws.sageMakerNotebook;
+
+import bio.terra.common.exception.ApiException;
+import bio.terra.common.exception.NotFoundException;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.common.utils.AwsUtils;
+import bio.terra.workspace.service.workspace.AwsCloudContextService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.services.sagemaker.model.NotebookInstanceStatus;
+
+public class StopAwsSageMakerNotebookStep implements Step {
+
+  private static final Logger logger = LoggerFactory.getLogger(StopAwsSageMakerNotebookStep.class);
+
+  private final ControlledAwsSageMakerNotebookResource resource;
+  private final AwsCloudContextService awsCloudContextService;
+  private final boolean resourceDeletion;
+
+  public StopAwsSageMakerNotebookStep(
+      ControlledAwsSageMakerNotebookResource resource,
+      AwsCloudContextService awsCloudContextService,
+      boolean resourceDeletion) {
+    this.resource = resource;
+    this.awsCloudContextService = awsCloudContextService;
+    this.resourceDeletion = resourceDeletion;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    AwsCredentialsProvider credentialsProvider =
+        AwsUtils.createWsmCredentialProvider(
+            awsCloudContextService.getRequiredAuthentication(),
+            awsCloudContextService.discoverEnvironment());
+
+    try {
+      NotebookInstanceStatus notebookStatus =
+          AwsUtils.getSageMakerNotebookStatus(credentialsProvider, resource);
+
+      switch (notebookStatus) {
+        case IN_SERVICE -> {
+          AwsUtils.stopSageMakerNotebook(credentialsProvider, resource);
+          AwsUtils.waitForSageMakerNotebookStatus(
+              credentialsProvider, resource, NotebookInstanceStatus.STOPPED);
+        }
+        case STOPPING -> AwsUtils.waitForSageMakerNotebookStatus(
+            credentialsProvider, resource, NotebookInstanceStatus.STOPPED);
+        case DELETING -> {
+          if (!resourceDeletion) {
+            throw new ApiException(
+                String.format(
+                    "AWS SageMaker Notebook resource %s, being deleted.",
+                    resource.getResourceId(), notebookStatus));
+          }
+          // else: being deleted
+        }
+        case PENDING, UPDATING, UNKNOWN_TO_SDK_VERSION -> throw new ApiException(
+            String.format(
+                "Cannot stop AWS SageMaker Notebook resource %s, status %s.",
+                resource.getResourceId(), notebookStatus));
+        case STOPPED, FAILED -> {
```

Comment:
Does `FAILED` mean the status call failed or the notebook instance is in a failed state? Is it possible that we can end up with orphaned notebooks?

Reply:
FAILED -> the notebook instance is in failed state i.e. not in service. Hence the deletion can go on if needed
No, this wont lead to orphaned notebooks.